### PR TITLE
Expose some HTTP server metrics like per-route request latency

### PR DESCRIPTION
**Pull request by charleskorn**
This PR adds some basic metrics for incoming HTTP requests, like per-route latency and status codes.

This uses the instrumentation from dskit, which required changing the HTTP router from a vanilla `net/http.ServeMux` to a `github.com/gorilla/mux.Router`. I'm not sure if we want to take a dependency on dskit here - open to feedback or other suggestions.

### CODE REVIEW

File: cmd/rollout-operator/tls_server.go [hunk 1]
```diff
@@ -9,31 +9,32 @@ import (
...
 
-func newTLSServer(port int, logger log.Logger, cert tlscert.Certificate) (*tlsServer, error) {
+func newTLSServer(port int, logger log.Logger, cert tlscert.Certificate, metrics *metrics) (*tlsServer, error) {
 	pair, err := tls.X509KeyPair(cert.Cert, cert.Key)
 	if err != nil {
 		return nil, fmt.Errorf("failed to load key pair: %v", err)
 	}
 
-	mux := http.NewServeMux()
+	m, handler := newInstrumentedRouter(metrics)
```
Comment (replay) [inline cmd/rollout-operator/tls_server.go:30]: we could consider currying these metrics with a label which identifies the protocol as http or https, but that can also be done in a future PR so I'll approve this one

**Review by replay — APPROVED** [review]

**Comment by pracucci** [comment]
> I'm not sure if we want to take a dependency on dskit here - open to feedback or other suggestions.

Adding dskit dependency looks fine to me. That's why we have a shared library for common stuff between our projects.

**Review by pracucci — APPROVED** [review]
LGTM. About Mauro's comment related to add scheme to metrics I don't have a strong opinion. In other projects we don't do it, so not adding it would keep consistency (unless we find out it's an important info).

File: cmd/rollout-operator/tls_server.go (hunk 1 above)
Comment (charleskorn) [reply cmd/rollout-operator/tls_server.go:30]: We register different routes on the HTTP and HTTPS servers, so I don't think this is necessary.

**Pull request merged**